lms/djangoapps/lti_provider/users.py

```diff
@@ -44,34 +47,39 @@ def authenticate_lti_user(request, lti_user_id, lti_consumer):
         switch_user(request, lti_user, lti_consumer)
 
 
-def create_lti_user(lti_user_id, lti_consumer):
+def create_lti_user(lti_user_id, lti_consumer, email=""):
     """
     Generate a new user on the edX platform with a random username and password,
     and associates that account with the LTI identity.
     """
     edx_password = str(uuid.uuid4())
 
-    created = False
-    while not created:
-        try:
-            edx_user_id = generate_random_edx_username()
-            edx_email = f"{edx_user_id}@{settings.LTI_USER_EMAIL_DOMAIN}"
-            with transaction.atomic():
-                edx_user = User.objects.create_user(
-                    username=edx_user_id,
-                    password=edx_password,
-                    email=edx_email,
-                )
-                # A profile is required if PREVENT_CONCURRENT_LOGINS flag is set.
-                # TODO: We could populate user information from the LTI launch here,
-                # but it's not necessary for our current uses.
-                edx_user_profile = UserProfile(user=edx_user)
-                edx_user_profile.save()
-            created = True
-        except IntegrityError:
-            # The random edx_user_id wasn't unique. Since 'created' is still
-            # False, we will retry with a different random ID.
-            pass
+    existing_user = User.objects.filter(email=email).first() if email else None
+
+    if existing_user:
+        edx_user = existing_user
+    else:
+        created = False
+        while not created:
+            try:
+                edx_user_id = generate_random_edx_username()
+                edx_email = f"{edx_user_id}@{settings.LTI_USER_EMAIL_DOMAIN}"
+                with transaction.atomic():
+                    edx_user = User.objects.create_user(
+                        username=edx_user_id,
+                        password=edx_password,
+                        email=edx_email,
+                    )
+                    # A profile is required if PREVENT_CONCURRENT_LOGINS flag is set.
+                    # TODO: We could populate user information from the LTI launch here,
+                    # but it's not necessary for our current uses.
+                    edx_user_profile = UserProfile(user=edx_user)
+                    edx_user_profile.save()
+                    created = True
+            except IntegrityError:
+                # The random edx_user_id wasn't unique. Since 'created' is still
+                # False, we will retry with a different random ID.
+                pass
```

Comment:
@Cup0fCoffee Because of the way you wrote this function, now a new use case jumped to my mind. What happens when we encounter an email in the LTI request, but the user doesn't have an account? The current implementation would create an anonymous user with a random ID. And even after they explicitly create an account on the platform, they would still end up using the anonymous account because the `authenticate_lti_user` function only queries the user with `lti_user_id` and `lti_consumer`. Which means, the auto-linking will only be possible for the specific users, who have an account on the Open edX platform and are using LTI for the first time using that specific LTI Consumer credentials. 

I am not sure what a good default here is, creating an account on behalf of someone else on the platform is probably a PII issue. So we can't create a new user with the `lis..email` every time we get an email. Which also means we this flag should be set only during the LtiConsumer creation, changing it later will not do the intended job of linking learners who are already associated with an anonymous account.

Ah! This is suddenly more complicated than before. :P

Reply:
After some thinking
* creating accounts for users with email is definitely a bad idea. Depending on the auth setup like SSO, this can have unintended consequences.
* allowing the flag to be changes back and forth will create the issue of having some users associated with their actual account and others with an anonymous account. This will cause confusion. So, the flag should only set when the `LtiConsumer` is created and can't be unset. If, for some reason, the auto-linking is no longer desired, then a new 
`LtiConsumer` should be created with the flag off.
* Finally, we will document that this auto-linking only works when the learners already have an account on the platform and can't be used to retrospectively link learners.

Reply:
@Cup0fCoffee I implemented your suggestion but, without defining a new function. I hope that's okay :)